### PR TITLE
Replace C binding layer with static typing hints

### PR DESCRIPTION
**Pull request by allisonkarlitskaya**

### CODE REVIEW

File: src/systemd_ctypes/librarywrapper.py [hunk 1]
```diff
@@ -15,98 +15,176 @@
...
+    NewType,
+    NoReturn,
+    Self,
+    Tuple,
+    TypeVar,
+)
+
+logger = logging.getLogger(__name__)
+
+if typing.TYPE_CHECKING:
```
Comment (martinpitt) [inline src/systemd_ctypes/librarywrapper.py:41]: Note to self: https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING

**Review by martinpitt — COMMENTED** [review]
This is really neat, thanks! All the tests are still red, but hopefully not too hard to salvage

File: pyproject.toml [hunk 2]
```diff
@@ -46,19 +46,22 @@ exclude_lines = [
...
 
 [testenv]
-sitepackages = true
+package = wheel
 deps=
-  pyflakes
   pytest-cov
   python-dbusmock
-commands =
-  python3 -m pyflakes src/systemd_ctypes test
```
Comment (martinpitt) [inline pyproject.toml:59]: This is the "pyproject.toml: overhaul tox config, again" commit, and it does not document dropping pyflakes at all. Was that an accident?

File: pyproject.toml [hunk 3]
```diff
@@ -16,10 +16,6 @@ Home = "https://github.com/allisonkarlitskaya/systemd_ctypes/"
 [tool.mypy]
 mypy_path = 'src'
 
-[tool.pytest.ini_options]
-pythonpath = ["src"]
-log_cli = true
```
Comment (martinpitt) [inline pyproject.toml:29]: This breaks running `pytest`. The commit message says so, but this feels unnecessary and evil. This makes debugging failures annoying, as it's yet another layer to peel off (aside from having to install yet another big thing into the test environment).

If you really insist on this, please update README, as it's no longer true.

File: src/systemd_ctypes/bus.py [hunk 4]
```diff
@@ -39,7 +38,7 @@ class BusError(Exception):
     :name: the 'code' of the error, like org.freedesktop.DBus.Error.UnknownMethod
     :message: a human-readable description of the error
     """
-    def __init__(self, name, message):
+    def __init__(self, name: str, message: str):
```
Comment (martinpitt) [inline src/systemd_ctypes/bus.py:41]: Introducing typing hints is very nice of course, but I notice that the tests don't actually run mypy. It fails a lot currently:

```
src/systemd_ctypes/bustypes.py:435: error: Dict entry 0 has incompatible type "Type[bool]": "Type[bool]"; expected "type": "BusType"  [dict-item]
src/systemd_ctypes/bustypes.py:436: error: Dict entry 1 has incompatible type "Type[bytes]": "Type[bytes]"; expected "type": "BusType"  [dict-item]
src/systemd_ctypes/bustypes.py:437: error: Dict entry 2 has incompatible type "Type[int]": "Type[int]"; expected "type": "BusType"  [dict-item]
src/systemd_ctypes/bustypes.py:438: error: Dict entry 3 has incompatible type "Type[str]": "Type[str]"; expected "type": "BusType"  [dict-item]
src/systemd_ctypes/bustypes.py:439: error: Dict entry 4 has incompatible type "Type[Variant]": "Type[Dict[Any, Any]]"; expected "type": "BusType"  [dict-item]
src/systemd_ctypes/bustypes.py:468: error: "type" has no attribute "__origin__"  [attr-defined]
src/systemd_ctypes/bustypes.py:469: error: "type" has no attribute "__args__"; maybe "__flags__"?  [attr-defined]
src/systemd_ctypes/event.py:49: error: "sd_event_source" has no attribute "wrapper"  [attr-defined]
src/systemd_ctypes/event.py:50: error: Too many arguments for "_add_inotify" of "sd_event"  [call-arg]
src/systemd_ctypes/event.py:50: error: Argument 3 to "_add_inotify" of "sd_event" has incompatible type "Event"; expected "_FuncPointer"  [arg-type]
src/systemd_ctypes/event.py:50: error: "sd_event_source" has no attribute "wrapper"  [attr-defined]
src/systemd_ctypes/event.py:50: error: Argument 4 to "_add_inotify" of "sd_event" has incompatible type "_FuncPointer"; expected "c_void_p"  [arg-type]
src/systemd_ctypes/event.py:113: error: Incompatible types in assignment (expression has type "Callable[[Coroutine[None, None, None], Optional[bool]], None]", variable has type "Callable[[Coroutine[Any, Any, _T], DefaultNamedArg(Optional[bool], 'debug')], _T]")  [assignment]
src/systemd_ctypes/event.py:115: error: Module has no attribute "_systemd_ctypes_polyfills"  [attr-defined]
src/systemd_ctypes/bus.py:76: error: "Reference[sd_bus_error]" has no attribute "contents"  [attr-defined]
src/systemd_ctypes/bus.py:155: error: Argument 2 to "send" of "sd_bus" has incompatible type "None"; expected "Reference[c_uint64]"  [arg-type]
src/systemd_ctypes/bus.py:361: error: Argument 4 to "_call_async" of "sd_bus" has incompatible type "None"; expected "c_void_p"  [arg-type]
src/systemd_ctypes/bus.py:517: error: Need type annotation for "members"  [var-annotated]
src/systemd_ctypes/bus.py:681: error: "org_freedesktop_DBus_Properties" has no attribute "_dbus_bus"  [attr-defined]
src/systemd_ctypes/bus.py:732: error: Argument 1 to "message_new_signal" of "Bus" has incompatible type "Optional[str]"; expected "str"  [arg-type]
test/test_basic.py:21: error: Skipping analyzing "dbusmock": module is installed, but missing library stubs or py.typed marker  [import]
test/test_basic.py:21: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
test/test_bustypes.py:231: error: Argument 1 to "message_new_method_call" of "Bus" has incompatible type "None"; expected "str"  [arg-type]
```

It's better than on main (which has 52 failures), but at some point these should be CI'ed. Are you still planning this, or want to keep it as a manual thing for now?

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks!

File: src/systemd_ctypes/bus.py (hunk 4 above)
Comment (allisonkarlitskaya) [reply src/systemd_ctypes/bus.py:41]: I think, in general, I want to figure out a better way to handle linting.  I've seen some people use different environments for this....

File: pyproject.toml (hunk 2 above)
Comment (allisonkarlitskaya) [reply pyproject.toml:59]: I merged this all into a mega-commit including introducing ruff and pycodestyle.

File: pyproject.toml (hunk 3 above)
Comment (allisonkarlitskaya) [reply pyproject.toml]: I brought this back.

**Pull request merged**